### PR TITLE
Add if constexpr macro and replace condition macro.

### PR DESCRIPTION
**Pull request by awulkiew**
This is a proposal adding conditional support of C++17 `if constexpr`. I modified only several files for now so we can start a discussion. If we have an agreement I'll modify other places.

My intention was to have a solution as close as possible to the original language feature and to keep `if` keyword so editors can highlight it as usual, i.e.:
```
if constexpr (static_condition)
```
vs
```
if BOOST_GEOMETRY_CONSTEXPR (static_condition)
```

It's a followup of https://github.com/boostorg/geometry/pull/1160

### CODE REVIEW

File: include/boost/geometry/util/constexpr.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#ifndef BOOST_NO_CXX17_IF_CONSTEXPR
+
+#define BOOST_GEOMETRY_CONSTEXPR(CONDITION) constexpr (CONDITION)
+
+#else
+
+#define BOOST_GEOMETRY_CONSTEXPR(CONDITION) (BOOST_GEOMETRY_CONDITION(CONDITION))
+
+#endif
+
```
Comment (barendgehrels) [inline include/boost/geometry/util/constexpr.hpp:25]: :+1:

File: include/boost/geometry/algorithms/detail/buffer/buffer_inserter.hpp [hunk 2]
```diff
@@ -969,7 +969,7 @@ inline void buffer_inserter(GeometryInput const& geometry_input, OutputIterator
                  robust_policy, strategies);
 
     collection.get_turns();
-    if (BOOST_GEOMETRY_CONDITION(areal))
+    if BOOST_GEOMETRY_CONSTEXPR (areal)
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/buffer/buffer_inserter.hpp:972]: I like it

File: include/boost/geometry/algorithms/detail/buffer/buffer_inserter.hpp [hunk 3]
```diff
@@ -1011,9 +1010,12 @@ inline void buffer_inserter(GeometryInput const& geometry_input, OutputIterator
...
 
-    if (BOOST_GEOMETRY_CONDITION(distance_strategy.negative() && areal))
+    if BOOST_GEOMETRY_CONSTEXPR (areal)
     {
-        collection.discard_nonintersecting_deflated_rings();
+        if (distance_strategy.negative())
+        {
+            collection.discard_nonintersecting_deflated_rings();
+        }
     }
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/buffer/buffer_inserter.hpp:1019]: Looks better than the original code

**Review by barendgehrels — APPROVED** [review]
I agree with this approach, thank you!

**Review by vissarion — APPROVED** [review]
Thanks! Looks great!

**Comment by awulkiew** [comment]
@vissarion Thanks! This wasn't finished though. I'll create another PR with the rest. :)

**Comment by vissarion** [comment]
> @vissarion Thanks! This wasn't finished though. I'll create another PR with the rest. :)

I know. I just merged it since there were the 1.83 and the enhancement labels and the beta is closing today. I guess there is no problem to have the rest of the support in 1.84.

**Pull request merged**